examples/vae_text/vae_train.py

```diff
@@ -0,0 +1,463 @@
+                hparams=config.dec_pos_emb_hparams)
+            # decoder
+            self.decoder = tx.modules.TransformerDecoder(
+                # tie word embedding with output layer
+                output_layer=self.decoder_w_embedder.embedding,
+                hparams=config.trans_hparams)
+        else:
+            raise NotImplementedError
+
+        self.decoder_initial_state_size = decoder_initial_state_size
+
+        self.connector_mlp = tx.modules.MLPTransformConnector(
+            config.latent_dims * 2,
+            linear_layer_dim=self.encoder.cell.hidden_size * 2)
+
+        self.mlp_linear_layer = nn.Linear(
+            config.batch_size,
+            tx.modules.connectors.sum_output_size(decoder_initial_state_size))
+
+    def forward(self, data_batch, kl_weight):
+        # encoder -> connector -> decoder
+
+        tmp_batch = []
+        for i, data in enumerate(data_batch["text_ids"]):
+            filter_data = data[data != 3]
+            delta = data.size(0) - filter_data.size(0)
+            tmp_batch.append(filter_data)
+            data_batch["length"][i] -= delta
+        text_ids = torch.stack(tmp_batch).to(device)
+
+        input_embed = self.encoder_w_embedder(text_ids)
+        output_w_embed = self.decoder_w_embedder(text_ids[:, :-1])
+        _, ecdr_states = self.encoder(
+            input_embed,
+            sequence_length=data_batch["length"])
+
+        if config.decoder_type == "lstm":
+            output_embed = output_w_embed
+
+        elif config.decoder_type == 'transformer':
+
+            batch_size = text_ids.size(0)
+            max_seq_len = text_ids.size(1) - 1
+            batch_max_seq_len = torch.ones(
+                batch_size, dtype=torch.long, device=device) * max_seq_len
+            output_p_embed = self.decoder_p_embedder(
+                sequence_length=batch_max_seq_len)
+            output_w_embed = output_w_embed * config.hidden_size ** 0.5
+            output_embed = output_w_embed + output_p_embed
+
+        else:
+            raise NotImplementedError
+        mean_logvar = self.connector_mlp(ecdr_states)
+        mean, logvar = torch.chunk(mean_logvar, 2, 1)
+        kl_loss = kl_dvg(mean, logvar)
+
+        dst = MultivariateNormalDiag(
+            loc=mean,
+            scale_diag=torch.exp(0.5 * logvar))
+
+        latent_z = dst.rsample()
+        dcdr_states = tx.modules.connectors.mlp_transform(
+                latent_z,
+                self.decoder_initial_state_size,
+                self.mlp_linear_layer)
+
+        # decoder
+        if config.decoder_type == "lstm":
+            # concat latent variable to input at every time step
+            latent_z = torch.unsqueeze(latent_z, 1)
+
+            latent_z = latent_z.repeat([1, output_embed.size(1), 1])
+            output_embed = torch.cat([output_embed, latent_z], 2)
+
+            train_helper = self.decoder.create_helper(
+                decoding_strategy='train_greedy')
```

Comment:
There's no need to create the train helper every time. It will be cached after it is created for the first time. Actually, since by default the `train_greedy` helper is used for training, there's no need to explicitly create it.

Reply:
Corrected, thanks for pointing out!